docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+   </td>
+  </tr>
+  <tr>
+   <td>Labels and Annotations
+   </td>
+   <td>Custom metadata that needs to be applied to the <code>TrainJob</code> resources: JobSet, Job, Pods.
+   </td>
+  </tr>
+  <tr>
+   <td><code>PodSpecOverrides</code>
+   </td>
+   <td>Custom overrides that are specific to the <code>TrainJob</code> and need to be applied to the
+    <code>TrainJob</code> resources. For example, the user identity. Usually, it is managed by
+    custom admission webhooks that inject data to the <code>TrainJob</code> after user creates it
+    via Python SDK or <code>kubectl</code>
+   </td>
+  </tr>
+</table>
+
+### Example of TrainJob
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: TrainJob
+metadata:
+  name: torch-ddp
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
+    name: torch-distributed-multi-node
+  trainerConfig:
+    image: docker.io/custom-training
+    command:
+      - torchrun train.py
+    numNodes: 5
+    resourcesPerNode:
+      requests:
+        nvidia.com/gpu: 2
+```
+
+The above command will be converted as follows:
+
+```bash
+torchrun --nnodes=5 --nproc-per-node=2 train.py
+```
+
+Additionally, the Kubeflow Training SDK allows to create the above `TrainJob` using the Python API:
+
+```python
+def train_func():
+    import torch
+    class Net(torch.nn.Module):
+        """Create the PyTorch Model"""
+        ...
+    model = Net()
+
+    # Attach model to the distributor
+    torch.distributed.init_process_group(backend="nccl")
+    model = torch.nn.parallel.DistributedDataParallel(model)
+
+    # Train model
+    model.train()
+
+# Use Kubeflow SDK to create TrainJob.
+from kubeflow.training import TrainingClient
+
+TrainingClient().train(
+    name="torch-ddp",
+    func=train_func,
+    num_nodes=5,
+    resources_per_node={"gpu": 2},
+)
+```
+
+### Example of LLM Fine-Tuning
+
+This example shows how to create `TrainJob` to fine-tune LLama 7b:
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: TrainJob
+metadata:
+  name: tune-llama-with-yelp
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
+    name: torch-tune-llama-7b
+  datasetConfig:
+    s3:
+      bucket: datasets
+      path: custom-datasets/yelp-review
+```
+
+### The Trainer Config API
+
+The `TrainerConfig` represents the APIs that data scientists can use to configure trainer settings:
+
+```golang
+type TrainerConfig struct {
+
+  // Docker image for the Trainer.
+  Image string `json:"image,omitempty"`
+
+  // Command for the training container.
+  // Validate that command contains torchrun or mpirun.
+  Command []string `json:"command,omitempty"`
+
+  // Args for the training container.
+  Args []string `json:"args,omitempty"`
+
+  // Env for the training container.
+  Env []corev1.EnvVar `json:"env,omitempty"`
+
+  // Number of training nodes.
+  NumNodes *int32 `json:"numNodes,omitempty"`
+
+  // Resource for each node.
+  ResourcesPerNode []corev1.resoruces `json:"resourcesPerNode,omitempty"`
```

Comment:
@vsoch Do you see a use-case when resources for MPI launcher should be changed frequently of every Job submission ?

Reply:
For the launcher, if it's just an extra pod/node that exists to send everything off (and the host isn't included in the set to do work, at least given the design of the MPI operator is like this) it could be relatively small in terms of resources. It probably wouldn't need that much changing between job submissions under these circumstances. If you have more of the Flux Operator design where the launcher is also included in the hostlist and does work, then you'd likely want it to have the same resources as the other nodes.

In terms of general resources for the job between submissions, that will definitely change, but I think you are just asking about the MPI Operator launcher. @alculquicondor can likely comment.

Reply:
In that case, we have this:
- When launcher works as Node with index 0, resources are the same for every Job replica.
- When launcher is a coordinator, resources will be pre-defined in the blueprint.

In that case, I don't think we should allow user to set resources for coordinator in the `TrainJob` API.

Reply:
I agree - and minimally since we understand these two designs, we can omit that specification of resources until there is reason to add it.